bug: Disallow spaces and reserved keywords in custom languages in Code Playground settings

nithyarajmudhaliyar opened the pull request:
This PR adds validation logic to the "Add a new code playground" form in the organization settings to prevent conflicts with Markdown syntax.

**Specifically, this change:**
* **Disallows spaces** in the language name, as these break code block parsing.
* **Disallows reserved keywords** (`latex`, `math`, `quote`, `spoiler`) that are used for other Markdown features.
* **Refactors validation** by moving logic into `check_pygments_language` and storing restricted keywords in the `RealmPlayground` model.

**Note:** Previously, this PR enforced lowercase. Per discussion on CZO, that enforcement has been removed. This PR now strictly focuses on validation (blocking invalid characters and keywords) without altering casing.

**Fixes:** #24044

**How changes were tested:**
Manually verified in the development environment:
1. Tried adding a language with spaces (e.g., `test space`); confirmed that a red error message appears.
2. Tried adding a reserved keyword (e.g., `math`); confirmed that a red error message appears ("Invalid characters...").
3. Tried adding a mixed-case language (e.g., `Rust`); confirmed that it saves successfully as `Rust` (preserving case).

**Screenshots and screen captures:**
<table>
<tr>
<th>Before</th>
<th>After</th>
</tr>
<tr>
<td>

https://github.com/user-attachments/assets/be723020-f41a-4b33-8b5d-52f8112cbad9

</td>
<td>


https://github.com/user-attachments/assets/53ce2a9f-0888-4d53-b57b-643a8e534366



</td>

</tr>
<tr>
<td>

https://github.com/user-attachments/assets/552747eb-8caa-4208-85da-05eb34c82b80


</td>
<td>


https://github.com/user-attachments/assets/450e221a-e595-4015-aae9-0092af21d9d1


</td>
</tr>
<table>


https://github.com/user-attachments/assets/3984297c-1a88-45e5-9177-8969f22a0f5b



<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on web/src/settings_playgrounds.ts:
```diff
@@ -117,9 +117,30 @@ function build_page(): void {
+                    $t_html({defaultMessage: "This is a special keyword and cannot be used."}),
+                    $playground_status,
+                );
+                return;
```
I guess we should just add these checks in API. And then we would need to handle it here.

sahil839 replied:
I think you can add backend validation in this PR and remove the frontend validation for now atleast.

Lowercase enforcement can also be done in backend.

nithyarajmudhaliyar commented:
@sahil839 Thanks for the guidance! I have updated the PR and addressed all the feedback. Moved the validation logic to backend and removed the frontend logic. I added a new test case `test_add_realm_playground_validation` to `zerver/tests/test_realm_playgrounds.py` to ensure the new backend logic is fully covered and verified.
All CI tests are passing. Ready for review!

sahil839 commented on zerver/views/realm_playgrounds.py:
```diff
@@ -49,9 +49,10 @@ def add_realm_playground(
     url_template: str,
 ) -> HttpResponse:
     pygments_language = pygments_language.strip()
+    pygments_language = pygments_language.lower()
```
Issue mentions converting to lower case only for custom languages.

nithyarajmudhaliyar replied:
@sahil839 I have a quick clarifying question regarding "custom languages" before I push the final fix.
My understanding is that any playground added via this "Add new code playground" form is technically treated as a "custom playground" configuration in the database—even if it is for a standard language like `Python` or `Rust`.
If I enforce lowercase here, it ensures that if a user types `Python`, it saves as python, which correctly matches the `` ```python `` markdown syntax so the "Run" button appears.
Does this align with your requirement to "enforce lower case only for custom languages," or should I be adding logic to treat standard languages differently?

sahil839 commented on zerver/views/realm_playgrounds.py:
```diff
@@ -49,9 +49,10 @@ def add_realm_playground(
     if " " in pygments_language:
         raise JsonableError(_("Language name cannot contain spaces."))
-    
+
```
Please fix such changes in the commit which added the code itself.

nithyarajmudhaliyar commented:
@sahil839 I've updated the PR to focus strictly on validation (blocking spaces and keywords).

I removed the lowercase enforcement to avoid potentially breaking standard languages (like `Rust` vs` rust`), as discussed on [CZO](https://chat.zulip.org/#narrow/channel/9-issues/topic/limiting.20code.20playground.20language.20name/with/2329725). I also updated the commit message to `Fixes part of #24044.` so the issue stays open for future discussion on casing.

Ready for review!

sahil839 commented on zerver/views/realm_playgrounds.py:
```diff
@@ -48,6 +48,12 @@ def add_realm_playground(
+        raise JsonableError(_("Language name cannot contain spaces."))
+    restricted_keywords = {"latex", "math", "quote", "spoiler"}
+    if pygments_language in restricted_keywords:
+        raise JsonableError(_("This is a special keyword and cannot be used."))
```
I think would be better to move this check in `check_pygments_language` defined above so that all validation is together at one place.

And space check can be done by updating the regex in `check_pygments_language`.

Also the `restricted_keywords` can be defined in `RealmPlayground` model, I think that would be a better place and is intutive place to find any such restrictions on a model field.

sahil839 commented on zerver/tests/test_realm_playgrounds.py:
```diff
@@ -88,6 +88,27 @@ def test_invalid_params(self) -> None:
+            "pygments_language": "rust lang",
+            "url_template": "https://example.com",
+        }
+        resp = self.api_post(iago, "/api/v1/realm/playgrounds", payload)
```
I think `result` would be a better variable name than `resp` and also we use `result` usually in other places as well.

nithyarajmudhaliyar commented:
@sahil839 I have updated the PR based on the feedback.
- Moved `RESTRICTED_KEYWORDS` into the `RealmPlayground` model.
- Moved all validation logic (including the keyword check) into `check_pygments_language`.
- Updated the regex to handle spaces automatically.
- Renamed `resp` to` result` in the tests.

Ready for Review!

sahil839 commented:
Can you update the screen recordings as well in the PR description as per the latest behavior? Code changes look good.

nithyarajmudhaliyar commented:
@alya The list of special characters (`_`,`+`,`-`,`.`, `/`, `#`) was chosen to support standard programming language identifiers and aliases used by Pygments.
 - `+` is required for languages like C++ ,
 - `#` is required for languages like C# and F# ,
 - `-` is required for Objective-C and Common-Lisp ,
 - `.` is required for VB.net .
 
 The goal was to allow valid language identifiers while blocking characters that cause Markdown parsing issues (like spaces) or conflicts (like keywords).
With that in mind, would you like me to proceed with the error message: `Language names can only contain letters, numbers, and the characters . - _ / # +`?

alya commented:
I guess I don't understand why the approach here is to whitelist characters, rather than just prohibiting the handful that are problematic. There are obscure languages with all sorts of funny character names.

sahil839 commented:
Code changes look good to me. Would be good to mention in the commit message about how you have updated the error message.

I am not sure if error message is clear enough like it is difficult to read what all characters are allowed. @alya any suggestions on that?

alya commented:
I'd change the error to `Failed: Invalid character in language: {character}`

- Indicate what field the error is about.
- I think the `'` around the character just make it harder to read.

I'd also leave the error banner showing until the user tries to submit again or navigates away.

alya commented:
This PR can be tagged for integration review when ready without further testing from me.

sahil839 commented on web/e2e-tests/realm-playground.test.ts:
```diff
@@ -14,7 +14,12 @@ async function _add_playground_and_return_status(page: Page, payload: Playground
+        if (el) {
+            el.style.display = "none";
+        }
+    }, admin_playground_status_selector);
```
I am not sure if this is correct. This hides the selector even when it is not hidden actuially till we submit the form.

sahil839 commented on web/e2e-tests/realm-playground.test.ts:
```diff
@@ -14,7 +14,6 @@ async function _add_playground_and_return_status(page: Page, payload: Playground
     await page.waitForSelector(".admin-playground-form", {visible: true});
     // Let's first ensure that the success/failure status from an earlier step has disappeared.
     const admin_playground_status_selector = "div#admin-playground-status";
-    await page.waitForSelector(admin_playground_status_selector, {hidden: true});
```
Also, I am not sure but this test can be flaky after changing the behavior of error message not hiding when checking the error message for the second case if it takes some time for server to complete the request as the error message is already visible and we do not have any way to check if the request was complete. Should we wait for the button to be re-enabled since that happens after the request is complete?
